cleanrl/ppo2_continuous_action.py

```diff
@@ -232,6 +258,31 @@ def get_logproba(self, x, actions):
         dist = Normal(action_mean, action_std)
         return dist.log_prob(actions).sum(1)
 
+    def get_entropies(self, x):
+        action_mean, action_logstd = self.forward(x)
+        action_std = torch.exp(action_logstd)
+        dist = Normal(action_mean, action_std)
+
+        # TODO: Discuss either using the logprobs instead of the entropy or the "pure" erntropy itself ?
+        sampled_actions = dist.rsample()
+        logprobs = dist.log_prob(sampled_actions).sum(1)
+        sampled_actions_probs = logprobs.exp()
+
+        # print( "# DEBUG: Actions logprobs shape")
+        # print( logprobs.shape)
+
+        entropies = - logprobs.exp() * logprobs
+
+        # print( "# DEBUG Entropy shape")
+        # print( entropy.shape)
+
+        # print( "# DEBUG: Manual entropy computation")
+        # print( - logprobs)
+        # print( "# DEBUG: Pytorch entropy computation")
+        # print( dist.entropy().sum(1))
+
+        return entropies
```

Comment:
The `get_entropies` function could be combined with
```python 
    def get_action(self, x):
        mean, logstd = self.forward(x)
        std = torch.exp(logstd)
        probs = Normal(mean, std)
        action = probs.sample()
        return action, probs.log_prob(action).sum(1)

    def get_logproba(self, x, actions):
        # Note: Converts actions to tensor
        if not isinstance(actions, torch.Tensor):
            actions = preprocess_obs_fn(actions)

        action_mean, action_logstd = self.forward(x)
        action_std = torch.exp(action_logstd)
        dist = Normal(action_mean, action_std)
        return dist.log_prob(actions).sum(1)
```

and get something like this:

```python
    def get_action(self, x, action=None):
        mean, logstd = self.forward(x)
        std = torch.exp(logstd)
        probs = Normal(mean, std)
        action = probs.sample() if not action
        return action, probs.log_prob(action).sum(1), dist.entropy().sum(1)
```

If you decided to go with Manual entropy computation, then do the entropy like in https://github.com/vwxyzjn/cleanrl/blob/master/cleanrl/ppo_continuous_action.py
```python
        newlogproba = pg.get_logproba(obs[:step+1], torch.Tensor(actions[:step+1]))
        # newvalues = vf.forward(obs[:step+1]).flatten() DO we generate a new values from the current policy?
        ratio =  torch.exp(newlogproba - torch.Tensor(logprobs[:step+1]))
        surrogate1 = ratio * torch.Tensor(advantages[:step+1])
        surrogate2 = ratio.clamp(1 - args.clip_coef, 1 + args.clip_coef) * torch.Tensor(advantages[:step+1])
        policy_loss = - torch.mean(torch.min(surrogate1, surrogate2))
        vf_loss = torch.mean((values[:step+1] - torch.Tensor(returns[:step+1])).pow(2))

        entropy_loss = torch.mean(torch.exp(newlogproba) * newlogproba)

        total_loss = policy_loss + args.vf_coef * vf_loss + args.ent_coef * entropy_loss
        optimizer.zero_grad()
        total_loss.backward(retain_graph=True)
```

Also, I don't quite see why `rsample()` is necessary here.

Reply:
Thanks for the suggestions.

I have run some experiments with both methods for the entropy computation and while the average results are mostly similar, the manual computed entropy seems better, as per the following report : [WANDB Report PPO2 Entropy Computation](https://www.reddit.com/r/MachineLearning/comments/f5kkxb/discussion_lessons_learned_from_my_failures_in/?utm_medium=android_app&utm_source=share).

Also moved the entropy computation to the network update section.